app/controllers/omniauth_callbacks_controller.rb

```diff
@@ -0,0 +1,49 @@
+class OmniauthCallbacksController < Devise::OmniauthCallbacksController
+  SocialProfile::PROVIDERS.each do |provider|
+    define_method(provider.to_s) do
+      begin
+        current_user ? connect_social_profile : handle_sign_in
+      rescue AuthVerificationPolicy::OauthError => e
+        redirect_to root_path, flash: { notice: e.to_s }
+      end
+    end
+  end
+
+  private
+
+  def connect_social_profile
+    OauthConnectOrganizer.new(auth, current_user).call
+    redirect_to edit_user_registration_path
+  end
+
+  def auth
+    request.env["omniauth.auth"]
+  end
+
+  def handle_sign_in
+    user = if auth_verified?
+      VerifiedAuthOrganizer.new(auth).user
+    else
+      UnverifiedAuthOrganizer.new(auth).user
+    end
+    sign_in_and_redirect user, event: :authentication
+  end
```

Comment:
lets's split it into two methods, handler and auth organizer.
First one will handle sign in, and second one will handle 2 cases of auth organizer
```ruby
def handle_sign_in
  user = auth_organizer.new(auth).user

  sign_in_and_redirect user, event: :authentication
end

def auth_organizer
  auth_verified? ? VerifiedAuthOrganizer : UnverifiedAuthOrganizer
end
```